_posts/2021-03-15-service-provider-interface.md

```diff
@@ -0,0 +1,201 @@
+Java provides many SPIs. Here is a list of some of them:
+- **ResourceBundleProvider**: Used by service providers to load resource bundles for named modules.
+- **CurrencyNameProvider**: Used by service providers that provide localized currency symbols and display names for the `Currency` class.
+- **Driver**: Used by service providers to load database drivers.
+
+Let's go further with the `Driver` and try to understand how the database drivers are loaded in our applications. For example, if we examine the Postgresql jar file we will find a folder called `META-INF/services` which contains a file named `java.sql.Driver`. This configuration file contains the name of the implementation class provided by Postgresql for the Driver interface, in this case: `org.postgresql.Driver`.
+
+We note the same thing with the MySQL driver: The `java.sql.Driver` located in `META-INF/services` contains `com.mysql.cj.jdbc.Driver` which is the implementation of the `Driver` interface of MySQL database.
+If the two drivers are loaded in the classpath, the `ServiceLoader` will read the implementation class names from each file, then calls `Class.forName` with the class names and then `newInstance` to create an instance of the implementation classes.
+
+### Example of an SPI Implementation
+Now that we have an understanding of the SPI concepts, let's create an example of an SPI and load providers using the `ServiceLoader` class.
+Let's say that we have a librarian who needs an application to check whether a book is available in his library or not when requested by customers.
+We can do this by defining a service represented by a class named `LibraryService` and a service provider interface called `Library`.
+The `LibraryService` provides a singleton `LibraryService` object. This object retrieves the book from `Library` providers.
+The library service client which is in our case the application that we are building gets an instance of this service, and the service will search, instantiate and use `Library` service providers.
+The application developers may in the first place use a standard list of books that can be available in all libraries. Another user who deals with computer science books may require another list of books for his library (another library provider). In this case, it would be better if the user can add the new library with the desired books to the existing application without modifying its core functionality. The new library will just be plugged into the application.
+
+### Implementation
+We start by creating a `maven` root project that will contain all our sub-modules. We will call it `library-service-provider-example`.
+The sub-modules will be:
+- *library-service-provider*: Contains the Service Provider Interface `Library` and the service class to load the providers.
+- *library-demo*: An application to put all together and create a working example.
+- *standard-library*: The provider for a standard library chosen by the developers.
```

Comment:
maybe rename `standard-library` to something that is recognizable as a book library...my first instinct when reading "standard library" is some programming library. Suggestion: `classics-library` (because it contains the book classics)